### PR TITLE
Optimize parts in background

### LINKED ISSUE #35836: Automatically optimize on special conditions
Clickhouse has special options "do_not_merge_across_partitions_select_final" for ReplaceMergeTree Engine, which allows make fast select on the "optimized" partitions. Unfortunate you have to run manually (crontab)  "OPTIMIZE TABLE ... PARTITION ... FINAL" . 

It will be great to have some options which will tell Clickhouse to compact older partitions on some conditions:

i.e.
```
CREATE TABLE select_final (t DateTime, x Int32, string String) ENGINE = ReplacingMergeTree() PARTITION BY toYYYYMMDD(t) ORDER BY (x, t)  SETTINGS optimize_partition_after = 3600
```
Clickhouse will check if a partition is older than 3600 seconds and has multiple parts and in such condition will do automatically optimization (compaction) 

P.S. we just discussed with @alexey-milovidov about this options :-)

### PR DESCRIPTION
**Pull request by antonio2368**
### Changelog category (leave one):
- Improvement

### Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
Merge parts if every part in the range is older than a certain threshold. The threshold can be set by using `min_age_to_force_merge_seconds`. This closes #35836.

This is continuation of #39550 by @fastio who implemented most of the logic.

> Information about CI checks: https://clickhouse.com/docs/en/development/continuous-integration/

### CODE REVIEW

**Comment by alexey-milovidov** [comment]
Alexey Milovidov
Why do we have to use two settings
auto_optimize_partition_interval_seconds
auto_optimize_partition_after_seconds
If we leave just one setting - a time passed after the newest part in partition appeared - should it suffice?


Alexey Milovidov
storage.auto_optimize_partition_task
Do we need a separate task? Can it be checked cheaply inside the normal selectPartsToMerge?


Alexey Milovidov
selectOnePartitionToOptimize
It can simply select parts to merge by this additional criteria inside MergeSelector, therefore maybe this function is unneeded.


Alexey Milovidov
TLDR I thought we can simplify the code much more, it will be just a few lines change.

**Comment by antonio2368** [comment]
@alexey-milovidov would something like this work?
First time going through this code and it seems correct but it would be nice to get a confirmation.

**Review by alexey-milovidov — APPROVED** [review]
Wow, the code becomes like 10x better!

**Comment by alexey-milovidov** [comment]
@antonio2368 I found that the logic is not quite as intended.
It was intended to select all parts within the partition if all of them are older than the threshold - if a whole partition is old.

Currently, it can select a subset of parts in the partition.

**Comment by antonio2368** [comment]
> Currently, it can select a subset of parts in the partition.

Yes, I wrote that in the description for the setting. I wasn't sure of the difference it would have, maybe the partition will merge too often if it's just a subset every time.
I will see what would be the easiest way to introduce a check for the entire partition.

**Comment by Algunenano** [comment]
There are use cases where merging parts more aggressively is desirable too. For example, when you have 2-3 largish parts of an AggregatingMergeTree they might stay unmerged for hours or days but you might want a more aggressive merge to speed up queries using those parts.

If this won't be possible with the changes we'll think of a proposal to cover that case.
